sdk/containerregistry/Azure.Containers.ContainerRegistry/src/Blob/ContainerRegistryBlobClient.cs

```diff
@@ -638,6 +680,100 @@ public virtual async Task<Response<DownloadBlobResult>> DownloadBlobAsync(string
             }
         }
 
+        private async Task<Response> DownloadBlobToInternalAsync(string digest, long blobLength, int maxChunkSize, Stream destination, bool async, CancellationToken cancellationToken)
+        {
+            if (blobLength <= maxChunkSize)
+            {
+                // Download in a single chunk
+                var downloadResult = async ?
+                    await DownloadBlobInternalAsync(digest, true, cancellationToken).ConfigureAwait(false) :
+                    DownloadBlobInternalAsync(digest, false, cancellationToken).EnsureCompleted();
+
+                if (async)
+                {
+                    await downloadResult.Value.Content.ToStream().CopyToAsync(destination).ConfigureAwait(false);
+                    await destination.FlushAsync(cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    downloadResult.Value.Content.ToStream().CopyTo(destination);
+                    destination.Flush();
+                }
+
+                return downloadResult.GetRawResponse();
+            }
+
+            // Download in multiple chunks.
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(maxChunkSize);
+            long chunkCount = 0;
+            long bytesDownloaded = 0;
+            using SHA256 sha256 = SHA256.Create();
+
+            try
+            {
+                Response result = null;
+
+                while (bytesDownloaded < blobLength)
```

Comment:
my understanding of [HTTP spec](https://www.rfc-editor.org/rfc/rfc7233#section-2.1) is that we don't need to know length ahead of time.

Service should respond to ranges if end is bigger than content-length

```
If a valid byte-range-set includes at least one byte-range-spec with
   a first-byte-pos that is less than the current length of the
   representation, or at least one suffix-byte-range-spec with a
   non-zero suffix-length, then the byte-range-set is satisfiable.
   Otherwise, the byte-range-set is unsatisfiable.

```
Service [should respond](https://www.rfc-editor.org/rfc/rfc7233#section-4.1) with `Content-Range` header that includes full blob size

e.g. `Content-Range: bytes 21010-47021/47022` (requested-range/size). 

I.m not sure what we do in other SDKs, but would be nice to avoid extra request (or getting size from user)

Reply:
One advantage to the expected blob size ahead of time is that is avoids a potentially large buffer allocation if the blob size is small.  But I agree it would be nice to avoid the extra request.

I guess if someone knows their blob will fit in memory, they call DownloadBlob and not DownloadBlobTo?

Reply:
> one advantage to the expected blob size ahead of time is that is avoids a potentially large buffer allocation if the blob size is small

i think we don't really need to allocate the buffer until we know what the blob size is, but this is micro-optimization


> I guess if someone knows their blob will fit in memory, they call DownloadBlob and not DownloadBlobTo?

 great point! I haven't even thought that we can explain the difference on the API level.